src/common/framework/tree.ts

```diff
@@ -1,82 +1,306 @@
+      const queryL1 = new TestQueryMultiFile(suite, entry.file);
+      const orderingL1 = compareQueries(queryL1, queryToLoad);
+      if (orderingL1 === Ordering.Unordered) {
+        // File path is not matched by this filter.
+        continue;
+      }
     }
 
-    if (!('g' in f.spec)) {
-      // This is a directory README.
+    if ('readme' in entry) {
+      // Entry is a readme that is an ancestor or descendant of the query.
+
+      // readmeSubtree is suite:a,b,*
+      const readmeSubtree: FilterResultSubtree<TestQueryMultiFile> = makeSubtreeForDirPath(
+        subtreeL0,
+        entry.file
+      );
+      readmeSubtree.description = entry.readme.trim();
       continue;
     }
+    // Entry is a spec file.
 
-    const [tRec] = log.record(f.id);
-    const fId = f.id.suite + ':' + f.id.path;
-    for (const t of f.spec.g.iterate(tRec)) {
-      let cases = tests;
-      for (const path of iteratePath(t.id.test, '~')) {
-        cases = getOrInsert(cases, fId + ':' + path);
+    const spec = await loader.importSpecFile(queryToLoad.suite, entry.file);
+    const description = spec.description.trim();
+    // subtreeL1 is suite:a,b:*
+    const subtreeL1: FilterResultSubtree<TestQueryMultiTest> = makeSubtreeForFilePath(
+      subtreeL0,
+      entry.file,
+      description,
+      checkCollapsible
+    );
+
+    // TODO: this is taking a tree, flattening it, and then unflattening it. Possibly redundant?
+    for (const t of spec.g.iterate()) {
+      {
+        const queryL3 = new TestQuerySingleCase(suite, entry.file, t.id.test, t.id.params);
+        const orderingL3 = compareQueries(queryL3, queryToLoad);
+        if (orderingL3 === Ordering.Unordered || orderingL3 === Ordering.StrictSuperset) {
+          // Case is not matched by this filter.
+          continue;
+        }
       }
 
-      const p = stringifyPublicParams(t.id.params);
-      cases.children!.set(fId + ':' + t.id.test + '=' + p, {
-        runCase: t,
-      });
+      // subtreeL2 is suite:a,b:c,d:*
+      const subtreeL2: FilterResultSubtree<TestQueryMultiCase> = makeSubtreeForTestPath(
+        subtreeL1,
+        t.id.test,
+        checkCollapsible
+      );
+
+      // Leaf for case is suite:a,b:c,d:x=1;y=2
+      makeLeafForCase(subtreeL2, t, checkCollapsible);
+
+      foundCase = true;
     }
   }
+  const tree = new FilterResultTree(subtreeL0);
+
+  for (const [i, sq] of subqueriesToExpandEntries) {
+    const seen = seenSubqueriesToExpand[i];
+    assert(seen, `subqueriesToExpand entry did not match anything: ${sq.toString()}`);
+  }
+  assert(foundCase, 'Query does not match any cases');
+
+  // TODO: Contains lots of single-child subtrees. Consider cleaning those up (as postprocess?).
+  return tree;
+}
+
+function makeTreeForSuite(suite: string): FilterResultSubtree<TestQueryMultiFile> {
+  return {
+    query: new TestQueryMultiFile(suite, []),
+    children: new Map(),
+    collapsible: false,
+  };
+}
+
+function makeSubtreeForDirPath(
```

Comment:
nit: addSubTreeForDirPath since we use getOrInsertSubtree. I was confused because I thought `tree` was const.